### PR TITLE
JsonLiteral retains type of literal

### PR DESCRIPTION
**Pull request by reitzig**
Here's a proposal for a `JsonLiteral` that knows which type of value the original JSON had (cf. #303).

I tried not to break its interface; users of the class need to add the wildcard type parameter `<*>`.

Does making `JsonLiteral` generic have consequences beyond what the tests cover? Can't say I have an overview over the actual object serialization parts of the library.

I can see potential changes to the interface but I'm not sure how far to go in an experimental library, so I stayed conservative.
On concrete thing: `JsonNull` could be a `JsonLiteral` now, couldn't it?

### CODE REVIEW

**Comment by reitzig** [comment]
This is a straight-up implementation of the standard now, which breaks quite a few tests.

 1. The tests seem to assume that output of `unquoted` can be parsed again. That _seems_ like relying on undocumented behaviour that contradicts the JSON spec -- or was it a conscious design decision?
 2. `nonstrict` is not supported fully. `NaN`, `[-]Infinity` are not parsed as numbers.
 3. Anything that I'm missing?

Assuming that non-strict JSON is a requirement, I think both parsing unquoted strings and non-strict mode can be done following the current approach; maybe `JsonReader` needs to get option flags.

Both at the same time, however, can not be parsed unambiguously: `NaN`, for example, might be a number (haha) or a string. While the deserialization consumers would know which conversion to use (based on the expected type), a stand-alone JSON parser (such as `Json.plain.parseJson)`) wouldn't know which literal type to choose.

Please advise as to what the expected/required behaviour is here, in particular for item 1 above.

**Comment by reitzig** [comment]
@sandwwraith and @qwwdfsad, are you waiting for a complete pull request before reviewing it, or are you having little time for it? Trying to decide whether to wait for your feedback or just continue to forge ahead.

**Comment by qwwdfsad** [comment]
Hi @reitzig 
I will review it tomorrow properly

File: runtime/common/src/main/kotlin/kotlinx/serialization/json/JsonElement.kt [hunk 1]
```diff
@@ -146,25 +146,12 @@ public sealed class JsonPrimitive : JsonElement() {
  * and `JsonLiteral("42", true)` will be serialized as `42` and `"42"` respectively.
  */
 @Serializable(JsonLiteralSerializer::class) // TODO val for body is workaround for plugin bug
-public class JsonLiteral internal constructor(val body: Any, public val isString: Boolean) : JsonPrimitive() {
+sealed class JsonLiteral<T> private constructor(val body: T) : JsonPrimitive() {
```
Comment (qwwdfsad) [inline runtime/common/src/main/kotlin/kotlinx/serialization/json/JsonElement.kt]: please do not remove explicit public modifier, it is crucial to have a proper understanding of what is and what is not public API (see [KEEP#45](https://github.com/Kotlin/KEEP/issues/45))

File: runtime/common/src/main/kotlin/kotlinx/serialization/json/JsonElement.kt [hunk 2]
```diff
@@ -146,25 +146,12 @@ public sealed class JsonPrimitive : JsonElement() {
...
+    @Deprecated("Check types instead")
+    public abstract val isString: Boolean
```
Comment (qwwdfsad) [inline runtime/common/src/main/kotlin/kotlinx/serialization/json/JsonElement.kt]: This property was not released, so you can safely remove it. 
Also, it should not be used in the implementation if it is deprecated

File: runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/JsonReader.kt [hunk 3]
```diff
@@ -76,12 +88,17 @@ internal val C2TC = ByteArray(CTC_MAX).apply {
...
+    initC2TC(TRUE.first(), TC_BOOL)
+    initC2TC(NUMBER_NEG, TC_NUMBER)
+    NUMBER_DIGITS.forEach { initC2TC(it, TC_NUMBER) }
 }
 
 // object instead of @SharedImmutable because there is mutual initialization in [initC2ESC]
 internal object EscapeCharMappings {
     @JvmField
-    public val ESCAPE_2_CHAR = CharArray(ESC2C_MAX)
+    val ESCAPE_2_CHAR = CharArray(ESC2C_MAX)
```
Comment (qwwdfsad) [inline runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/JsonReader.kt]: ditto`public`

File: runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/JsonReader.kt [hunk 4]
```diff
@@ -37,6 +39,14 @@ internal const val STRING_ESC = '\\'
 internal const val INVALID = 0.toChar()
 internal const val UNICODE_ESC = 'u'
 
+internal const val NUMBER_NEG = '-'
+internal const val NUMBER_POS = '+'
+internal const val NUMBER_SEP = '.'
+internal const val NUMBER_EXP = 'e'
+internal val NUMBER_DIGITS = arrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9') // TODO: ?
```
Comment (qwwdfsad) [inline runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/JsonReader.kt]: It is better to use `CharArray` instead of `Array<Char>` for the performance sake

**Comment by qwwdfsad** [comment]
>nonstrict is not supported fully. NaN, [-]Infinity are not parsed as numbers.

I will fix it separately, thanks for catching it

File: runtime/common/src/main/kotlin/kotlinx/serialization/json/JsonElement.kt (hunk 2 above)
Comment (sandwwraith) [reply runtime/common/src/main/kotlin/kotlinx/serialization/json/JsonElement.kt]: it is `public` in master now

**Comment by sandwwraith** [comment]
I also doubt that this feature worth adding a generic parameter to `JsonLiteral` – see, in most of the places where it is used you had to erase this information with `<*>` because you simply do not care. It is also not a 'minimal breakage'. So an additional property with some representation of primitive kind and possibly some converter functions look more favorable to me.

**Comment by diegomontoya** [comment]
I also do not believe we actually need a generic JsonLiteral<T>. To FIX "10" vs 10 issue which is very realistic, we just need to implement a isQuoted property and expose it via public API.

**Comment by reitzig** [comment]
Thanks for the feedback! It's been a while, and I haven't really gotten around to this.

Tl;dr: based on the comments this PR received, it's become clear to me that the library was never intended to provide a cross-platform JSON parser as such. I went at this with wrong expectations; closing.

---

> There is a lot of heavy-lifting in JsonParser/JsonReader just to properly support this small feature. Moreover, due to this feature, regular JSON parser will experience a notable performance hit.

What makes you say that? There's a bit more code, but aside from that one, easily fixed `Array<Char>` usage aside, the only bit that changed the actual performance is the [number literal check](https://github.com/Kotlin/kotlinx.serialization/blob/f5d9b7c950780fcaadcd3a2b8cfdd989432a7791/runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/JsonReader.kt#L325) that could be circumvented if it's a bottleneck.
Note how even that change contributes something that's IMHO worthwhile (required, even): it checks for correct number literals at _parse time_, not when consuming the parse result as before (IIRC).

> The public API surface can (has to) be reduced to a single class change: JsonLiteral.

> I also doubt that this feature worth adding a generic parameter to JsonLiteral – see, in most of the places where it is used you had to erase this information with <*> because you simply do not care.

That was a kind-of uncomfortable compromise. I _actually_ wanted to remove `JsonLiteral` in favor of what then became its subtypes, but clearly _that_ would have been a "very breaking" change. Adding a type parameter technically breaks usages, but those are easily fixed by adding `<*>`, as you noted.

The whole point of this PR is to enable _new_ types of usage; of course, those (internal) usages that already exist didn't care about literal types, for better or worse, so they go with `<*>`. The schema validator I was working on would have relied heavily on the full types!

(FWIW, the "has to" in the first quote feels off to me, talking about an experimental library.)

> Keep API small and simple

Agreed. But those proposals are exactly the opposite, aren't they:

> So an additional property with some representation of primitive kind and possibly some converter functions look more favorable to me.

> To FIX "10" vs 10 issue which is very realistic, we just need to implement a isQuoted property and expose it via public API and let the end-user deal with it.

In summary, I still don't understand why the parser was written the way it was (it doesn't conform to the JSON specification)

**Comment by qwwdfsad** [comment]
@reitzig 

> it's become clear to me that the library was never intended to provide a cross-platform JSON parser as such

It's obviously not true, could you please elaborate on what made you think that way?

>FWIW, the "has to" in the first quote feels off to me, talking about an experimental library

Yes, the library is still experimental, but we are working hard towards its stabilization. It includes stabilization of runtime features such as JSON, Kotlin feature-completeness (objects, sealed classes, inline classes, delegates etc.) and re-evaluation of design-decisions in the code-generating plugin, including its architecture and code-generation strategy (to make it future-proof and to still have opportunities for optimizations and to change code-generation strategy even when serialization is stable). Thus now we are much more mindful about each new public signature or semantic change.

>I still don't understand why the parser was written the way it was (it doesn't conform to the JSON specification)

It was indeed written as an experiment in the first place, thus the code quality. But again, we are working towards changing that (e.g. #497). Now it should **mostly** conform the spec modulo some corners I don't really want to explore. I would even say that we are on par with GSON when talking about spec conformance.

>What makes you say that? There's a bit more code, but aside from that one, easily fixed Array<Char> usage aside, the only bit that changed the actual performance is the number literal check that could be circumvented if it's a bottleneck.

A few points here:
- More token classes to check for/parse. It may be insignificant though, but it's hard to tell without measuring it first.
- Double (and other types as well, double is just the biggest contributor) parsing when reading `JsonElement`, even if one doesn't need a precise type.
- `nextNumber` itself (that indeed can be optimized)

It's arguable whether it is really a blocker, but just something one should be aware about.


Regarding the change in general:

I realize this feature is important to you, but unfortunately, it does not mean that we should include this feature. And it doesn't mean that we are not interested in development of the library, being feature-rich or being contribution-friendly either! It's all about compromises (often mixed with personal preferences regarding API shape) that sometimes do not match your expectations.

For example, you are developing schema-validation. We were thinking about it as well and decided that it should work via `SerialDescriptor` hierarchy (very simplified example: https://github.com/Kotlin/kotlinx.serialization/blob/master/examples/example-visitors/src/test/kotlin/kotlinx/serialization/schema/JsonSchemaTest.kt) instead of relying on raw `JsonElement` types. Maybe this is what suits your needs instead of typed `JsonElement`?

**Comment by reitzig** [comment]
> It's obviously not true, could you please elaborate on what made you think that way?

IMHO, a general-purpose JSON parser would focus being conformant to the standard(s). I would expect the same library to be able to generate identical JSON (up to formatting) from its internal representation, without external information; `marshal(unmarshal(x) == x` and `unmarshal(marshal(x)) == x` seem like reasonable requirements in that sense.

Here, the priority lies on one specific use-case -- object/data serialization -- and efficiency. The assumption that separate type information is available (from a Kotlin class, say) seems inherently baked into the approach.
(This is speaking from what I say back when I created the PR. If things have changed, never mind.)

That's all fair -- as you say, different goals mean different trade-offs. It just meant that I saw that this library was not intended to serve the purpose I expected, so instead of trying to shoehorn it I had to go a different route.
